### PR TITLE
Point To Point Cut Tool

### PR DESCRIPTION
**Pull request by lopezt-unity**
### Purpose of this PR

New feature: Adding the point-to-point cut tool.
Using this tool the user can add a cut in the face of a PB Mesh, creating a division of the original face in different sub-faces.
cutFinal

This tool is based on `EditorTool`, so it is made to be used with Unity 2019.4 and above.

### Comments to Reviewers

[List known issues, planned work, provide any extra context for your code.]

### CODE REVIEW

**Comment by gabrielw-us** [comment]
Awesome 👍 

I noticed that edges do not highlight unless their face is visible, is this on purpose? Feels unexpected to me.

**Comment by gabrielw-us** [comment]
Also a bit unsure about the ESC key action, and "commit" result when path is invalid. Both times, very easy to lose complex cut, and no way to retrieve it, no warning this will happen. Also, works differently than Polyshape. I'll investigate! :)

**Review by JoelFortin — APPROVED** [review]
Peer-review with Thomas ... all the problems found in the first pass was fixed and looks pretty solid to me right now. Nice Work !!!

File: Editor/EditorCore/CutTool.cs
Comment (karl-) [inline Editor/EditorCore/CutTool.cs:76]: should we reuse any of the existing `Handles` pref colors here? cc @gabrielw-us

File: Editor/EditorCore/CutTool.cs
Comment (karl-) [inline Editor/EditorCore/CutTool.cs:98]: why are these all internal? who is accessing these outside of the class?

File: Editor/EditorCore/CutTool.cs
Comment (karl-) [reply Editor/EditorCore/CutTool.cs:76]: oh I see, they do later on in the ctor

File: Editor/EditorCore/CutTool.cs
Comment (karl-) [reply Editor/EditorCore/CutTool.cs:41]: would it be so that this tool continues to function correctly if a domain reload occurs during operation?

File: Editor/EditorCore/CutTool.cs
Comment (karl-) [inline Editor/EditorCore/CutTool.cs:191]: why is this done as a delayed call?

File: Editor/EditorCore/CutTool.cs
Comment (karl-) [inline Editor/EditorCore/CutTool.cs:278]: use `EditorGUIUtility.TrContent` for translation

File: Editor/EditorCore/CutTool.cs
Comment (karl-) [inline Editor/EditorCore/CutTool.cs:488]: if you're not adding or removing vertices you can save an alloc here by using `m_Mesh.positionsInternal`

File: Editor/EditorCore/CutTool.cs
Comment (karl-) [inline Editor/EditorCore/CutTool.cs:1172]: Use `EditorToolManager.RestorePreviousTool();`

File: Editor/EditorCore/CutTool.cs
Comment (karl-) [inline Editor/EditorCore/CutTool.cs:1247]: it would be nice to avoid the `ToArray` and linq calls here if we're running this on `EventType.Repaint`

File: Editor/EditorCore/CutTool.cs
Comment (karl-) [inline Editor/EditorCore/CutTool.cs:1257]: same note as above, can we reduce the amount of allocations by adding the appropriate overloads to `EditorHandleDrawing`?

File: Editor/EditorCore/CutTool.cs
Comment (karl-) [inline Editor/EditorCore/CutTool.cs:1294]: if this function is just reading the vertex data use `m_Mesh.positionsInternal` to skip the array copy.

File: Editor/EditorCore/PolyShapeEditor.cs
Comment (karl-) [inline Editor/EditorCore/PolyShapeEditor.cs:756]: what's going on here?

File: Runtime/Core/SubdivideOnVertex.cs
Comment (karl-) [inline Runtime/Core/SubdivideOnVertex.cs:12]: let's talk about this approach, not sure about using a monobehaviour here

**Review by karl- — CHANGES_REQUESTED** [review]

**Comment by karl-** [comment]
Screen Shot 2020-10-07 at 4 21 28 PM (2)

on macos the cursor size is incorrect

File: Editor/EditorCore/CutTool.cs
Comment (lopezt-unity) [reply Editor/EditorCore/CutTool.cs:41]: When it's not serializable the undo does not seem to work properly as positions from the cut are not serialized. Then Undo  does not remove the last position from the cut if the structure is not serializable.

File: Editor/EditorCore/CutTool.cs
Comment (lopezt-unity) [reply Editor/EditorCore/CutTool.cs:55]: only getter is needed, same for the type, the set are only done using the constructor.
This can be modified if needed.

File: Editor/EditorCore/CutTool.cs
Comment (lopezt-unity) [reply Editor/EditorCore/CutTool.cs:98]: The test class CutToolTest.cs is :) (Yeah, I did try to make a test class for this tool :D )

File: Editor/EditorCore/CutTool.cs
Comment (lopezt-unity) [reply Editor/EditorCore/CutTool.cs:191]: I think it was for a purpose that is not valid anymore => I removed it

File: Editor/EditorCore/CutTool.cs
Comment (lopezt-unity) [reply Editor/EditorCore/CutTool.cs:278]: What is the difference between this method and L10n.Tr method?

File: Runtime/Core/SubdivideOnVertex.cs
Comment (lopezt-unity) [reply Runtime/Core/SubdivideOnVertex.cs:12]: Oops, this was only a test functionality I did when starting in the team, I'm removing this class

File: Editor/EditorCore/PolyShapeEditor.cs
Comment (lopezt-unity) [reply Editor/EditorCore/PolyShapeEditor.cs:756]: > on macos the cursor size is incorrect
> 
> also to note - the cursor images are 512^2, which is way larger than they need to be in the first place. importing icons is the main cause for lengthy package import times.

The cursor images are now 64x64, feel free to modify the aspect on MacOS, I am on Windows and don't really know how to handle that in Mac :p

Comment (lopezt-unity) [reply Editor/EditorCore/PolyShapeEditor.cs]: This is not valid anymore after the merge with the new polyshape as editortool

**Comment by lopezt-unity** [comment]
> Screen Shot 2020-10-07 at 4 21 28 PM (2)
> 
> on macos the cursor size is incorrect
> 
> also to note - the cursor images are 512^2, which is way larger than they need to be in the first place. importing icons is the main cause for lengthy package import times.

Icons are now 64x64 pixels, which should be better.
@karl- could you correct the icon display size problem on MacOS, as I don't have one?

File: Editor/EditorCore/PolyShapeTool.cs
Comment (karl-) [inline Editor/EditorCore/PolyShapeTool.cs:764]: This might not be necessary anymore, I'm not sure.

File: Runtime/Core/ProBuilderMesh.cs
Comment (karl-) [inline Runtime/Core/ProBuilderMesh.cs:473]: xml docs don't match the function signature

File: Runtime/MeshOperations/QuadUtility.cs
Comment (karl-) [inline Runtime/MeshOperations/QuadUtility.cs:6]: nice! it's good to see this abstracted

File: Tests/Editor/Geometry/CutToolTest.cs
Comment (karl-) [inline Tests/Editor/Geometry/CutToolTest.cs:29]: Can use `Assume` when doing test state setup

**Review by karl- — APPROVED** [review]

File: Editor/EditorCore/PolyShapeTool.cs
Comment (lopezt-unity) [reply Editor/EditorCore/PolyShapeTool.cs:764]: This is not needed for the tool itself (move/rotate/scale) but is indeed needed to retrieve the previous PB mode (face/edge/vertex)

**Comment by lopezt-unity** [comment]
@karl- Could you have a look to the cursor icon issue on iOS by any chance?

**Comment by karl-** [comment]
> @karl- Could you have a look to the cursor icon issue on iOS by any chance?

@lopezt-unity I haven't yet. I don't want to block the PR for that though, let's just merge this and open a bug to fix it

**Pull request merged**
